rfcs/2023-rsp-testing-utils.md

```diff
@@ -0,0 +1,113 @@
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License. -->
+
+- Start Date: 2023/7/27
+- RFC PR: exploration PRs: https://github.com/adobe/react-spectrum/pull/4836, https://github.com/adobe/react-spectrum/pull/4837
+- Authors: Daniel Lu
+
+# Improving React Spectrum test writing experience
+
+## Summary
+
+This RFC proposes a three part improvement to the test writing experience for consumers of React Spectrum/Aria Components. It will cover possible additions to our documentation for test writing education and gotchas, discuss a list of low level helper/utility function to expose that can aid in test setup/interactions, and propose a new set of high level utility classes that aim to simplify writing various common component interactions during testing.
+
+## Motivation
+
+As both the adoption of React Spectrum and the complexity of the components offered has grown, various testing pain points have surfaced, both from within the maintaining team and from consumers of the library itself. For a outside consumer writing tests that run in a non-browser environment, it is hard to know what exactly needs to be mocked to make certain components (e.g. virtualized components) even render properly. They may not be familiar with the internal structure of a component and how best to target/interact those nodes, that a component has a transition or requestAnimationFrame internally and thus needs a equivalent wait time, or even what events need to be fired to simulate common operations like drag and drop. As a result, many people often have to directly reference the tests within the library itself or dig into the component's internals when they run into any test issues. This can be very time consuming and frustrating for someone trying to write a simple interaction test, and adds unnecessary friction for new adopters.
+
+These are non-trivial issues for maintainers of the library and have resulted in various internal testing utils and patterns, but things like common component interactions are typically duplicated across test files instead of being centralized. Explicit documentation of various testing gotchas doesn't really exist and has become internalized knowledge instead.
+
+We hope that documenting these testing gotchas externally will reduce the test debugging frustration, and that exposing testing utilities for common test setup and component interactions will simplify the overall test writing experience, leading towards easier adoption.
+
+
+## Detailed Design
+
+As mentioned earlier, there are three proposed parts to this initiative:
+
+1. Documentation of testing gotchas/common issues
```

Comment:
I think this would be really helpful. Maybe the React Spectrum and React Aria sections of the docs can have their own separate testing pages.